### PR TITLE
Running requests in background

### PR DESCRIPTION
**Pull request by abidlabs**
While traveling to Pakistan, I had the experience of trying to use Gradio with weak Wifi and... it's extremely slow to launch even a basic demo (took about 11 seconds to launch a hello world demo!). I decided to try to find opportunities to speed up the library and found some low hanging fruit, primarily converting some of our telemetry requests to run on background threads. I considered using the `AsyncRequest` class but we can't easily run these requests in an async context, so I used threading instead. 

**Before**:

```
3406290 function calls (3342711 primitive calls) in 11.098 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      186    0.010    0.000   20.535    0.110 __init__.py:1(<module>)
   1900/1    0.027    0.000   11.102   11.102 {built-in method builtins.exec}
        1    0.000    0.000   11.102   11.102 run.py:1(<module>)
   1956/1    0.037    0.000    9.315    9.315 <frozen importlib._bootstrap>:1002(_find_and_load)
   1949/1    0.019    0.000    9.315    9.315 <frozen importlib._bootstrap>:967(_find_and_load_unlocked)
   1852/1    0.020    0.000    9.314    9.314 <frozen importlib._bootstrap>:659(_load_unlocked)
   1703/1    0.014    0.000    9.314    9.314 <frozen importlib._bootstrap_external>:849(exec_module)
   2361/1    0.005    0.000    9.313    9.313 <frozen importlib._bootstrap>:220(_call_with_frames_removed)
   770/75    0.006    0.000    9.116    0.122 {built-in method builtins.__import__}
        1    0.000    0.000    7.791    7.791 components.py:1(<module>)
       21    0.001    0.000    7.535    0.359 __init__.py:2(<module>)
 1784/757    0.007    0.000    7.119    0.009 <frozen importlib._bootstrap>:1033(_handle_fromlist)
        1    0.000    0.000    3.202    3.202 processing_utils.py:1(<module>)
        4    0.000    0.000    3.001    0.750 api.py:16(request)
        4    0.000    0.000    2.992    0.748 sessions.py:463(request)
        4    0.000    0.000    2.975    0.744 sessions.py:614(send)
        4    0.000    0.000    2.972    0.743 adapters.py:394(send)
        4    0.000    0.000    2.966    0.742 connectionpool.py:499(urlopen)
        4    0.000    0.000    2.964    0.741 connectionpool.py:356(_make_request)
        3    0.000    0.000    2.933    0.978 api.py:64(get)
       24    0.002    0.000    2.702    0.113 utils.py:1(<module>)
        2    0.000    0.000    2.061    1.031 connectionpool.py:970(_validate_conn)
        2    0.000    0.000    2.061    1.031 connection.py:307(connect)
     1703    0.048    0.000    2.022    0.001 <frozen importlib._bootstrap_external>:921(get_code)
5417/5252    0.189    0.000    1.621    0.000 {built-in method builtins.__build_class__}
        2    0.000    0.000    1.618    0.809 blocks.py:1(<module>)
        1    0.000    0.000    1.521    1.521 networking.py:1(<module>)
        1    0.000    0.000    1.503    1.503 ipython_ext.py:1(<module>)
        1    0.000    0.000    1.462    1.462 routes.py:1(<module>)
        1    0.000    0.000    1.451    1.451 core.py:4(<module>)
        3    0.000    0.000    1.448    0.483 strings.py:1(<module>)
        8    0.001    0.000    1.367    0.171 core.py:1(<module>)
     1710    0.021    0.000    1.358    0.001 <frozen importlib._bootstrap_external>:1041(get_data)
        2    0.000    0.000    1.357    0.679 ssl_.py:308(ssl_wrap_socket)
        1    0.000    0.000    1.357    1.357 embed.py:2(<module>)
        1    0.000    0.000    1.321    1.321 blocks.py:446(__init__)
        1    0.000    0.000    1.319    1.319 utils.py:87(get_local_ip_address)```
```

**After**:

```
3402125 function calls (3338628 primitive calls) in 6.907 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      186    0.006    0.000   13.462    0.072 __init__.py:1(<module>)
   1900/1    0.017    0.000    6.910    6.910 {built-in method builtins.exec}
        1    0.000    0.000    6.910    6.910 run.py:1(<module>)
   1956/1    0.023    0.000    5.855    5.855 <frozen importlib._bootstrap>:1002(_find_and_load)
   1949/1    0.013    0.000    5.855    5.855 <frozen importlib._bootstrap>:967(_find_and_load_unlocked)
   1852/1    0.013    0.000    5.854    5.854 <frozen importlib._bootstrap>:659(_load_unlocked)
   1703/1    0.009    0.000    5.854    5.854 <frozen importlib._bootstrap_external>:849(exec_module)
   2361/1    0.003    0.000    5.853    5.853 <frozen importlib._bootstrap>:220(_call_with_frames_removed)
   769/75    0.003    0.000    5.647    0.075 {built-in method builtins.__import__}
        1    0.000    0.000    5.221    5.221 components.py:1(<module>)
       21    0.001    0.000    5.080    0.242 __init__.py:2(<module>)
 1784/757    0.004    0.000    4.455    0.006 <frozen importlib._bootstrap>:1033(_handle_fromlist)
        2    0.000    0.000    1.564    0.782 blocks.py:1(<module>)
        1    0.000    0.000    1.468    1.468 networking.py:1(<module>)
        1    0.000    0.000    1.421    1.421 routes.py:1(<module>)
     1703    0.031    0.000    1.348    0.001 <frozen importlib._bootstrap_external>:921(get_code)
5417/5252    0.133    0.000    1.222    0.000 {built-in method builtins.__build_class__}
        1    0.000    0.000    1.185    1.185 processing_utils.py:1(<module>)
        1    0.000    0.000    1.066    1.066 core.py:4(<module>)
        8    0.001    0.000    0.981    0.123 core.py:1(<module>)
     4379    0.066    0.000    0.926    0.000 <frozen importlib._bootstrap_external>:1505(find_spec)
        3    0.000    0.000    0.920    0.307 api.py:16(request)
        3    0.000    0.000    0.914    0.305 sessions.py:463(request)
        3    0.000    0.000    0.906    0.302 sessions.py:614(send)
        3    0.000    0.000    0.905    0.302 adapters.py:394(send)
     1710    0.012    0.000    0.903    0.001 <frozen importlib._bootstrap_external>:1041(get_data)
        3    0.000    0.000    0.903    0.301 connectionpool.py:499(urlopen)
        3    0.000    0.000    0.901    0.300 connectionpool.py:356(_make_request)
        2    0.000    0.000    0.868    0.434 api.py:64(get)
        1    0.000    0.000    0.861    0.861 blocks.py:446(__init__)
        1    0.000    0.000    0.860    0.860 utils.py:102(initiated_analytics)
        2    0.000    0.000    0.859    0.430 utils.py:87(get_local_ip_address
```

Closes: #2052 (in that all the potentially network requests that block / slow down launch now run in short-lived background threads instead)

### CODE REVIEW

**Comment by abidlabs** [comment]
I'm curious what the performance difference is on faster wifi. You can run this command to profile a demo:

```bash
python -m cProfile  -s time -o profile.txt .\demo\blocks_essay\run.py
```

File: gradio/utils.py [hunk 1]
```diff
@@ -84,58 +85,83 @@ def version_check():
...
 def get_local_ip_address() -> str:
-    """Gets the public IP address or returns the string "No internet connection" if unable to obtain it."""
-    try:
-        ip_address = requests.get(
-            "https://checkip.amazonaws.com/", timeout=3
-        ).text.strip()
-    except (requests.ConnectionError, requests.exceptions.ReadTimeout):
-        ip_address = "No internet connection"
+    """Gets the public IP address or returns the string "No internet connection" if unable to obtain it. Does not make a new request if the IP address has already been obtained."""
+    if Context.ip_address is None:
```
Comment (freddyaboulton) [inline gradio/utils.py:89]: What's the performance of this approach relative to keeping `get_local_ip_address` the same and then running all subsequent telemetry requests in separate threads?

Wondering if we can avoid using the `Context` as a cache for ip address

Comment (abidlabs) [reply gradio/utils.py:89]: IIRC it shaved about 1 second for me. Is there a reason this is a bad idea? The reason I did it this way was to allow `get_local_ip_address` to run in a background thread but only have it run once per session

Comment (freddyaboulton) [reply gradio/utils.py:89]: Don't think it's a bad idea but if I'm reading the current code correctly, we also only call `get_ip_address` once and then just store it in the blocks state so seems like we're already caching it. So using the `Context` as a cache seems like a more complicated way to do it.

Comment (abidlabs) [reply gradio/utils.py:89]: The problem with the original approach is doesn’t easily adapt to run in a background thread. A (secondary) advantage of this née approach is that its faster when creating multiple Interfaces/Blocks in a jupyter notebook where a single Context is shared

**Review by freddyaboulton — COMMENTED** [review]
Getting about 2x speedup in slowish Rome coffee shop wifi!

**Review by freddyaboulton — APPROVED** [review]

**Comment by abidlabs** [comment]
Sweet thanks @freddyaboulton!

**Comment by abidlabs** [comment]
Fixed tests and confirmed that analytics still show up on Segment: 

image

**Comment by FarukOzderim** [comment]
Hello, @abidlabs @freddyaboulton!

WDYT about adding these background requests [Startup-events].(https://github.com/gradio-app/gradio/blob/7fc3e750b04290c2066758f8697c575cef3f6441/gradio/routes.py#L475) 

1. It would be more tidy and structured.
2. They could run as async requests in the background, having a lighter weight on startup.

**Comment by abidlabs** [comment]
But some of these (e.g. the initiated analytics) happen before the Interface/Blocks is launched (they happen when the Interface object is created) so I don't think this would work @FarukOzderim?

**Comment by FarukOzderim** [comment]
Hmm, to be sure, they are not needed immediately right?
Then all of the prior calls could be stored inside a list which will be called in startup.

**Comment by abidlabs** [comment]
Hmm but what if the app is never launched? The purpose of the _initiated analytics_ was to log an event when the `Interface` object is created so that we could see if there are certain interfaces that users are not launching (e.g. using as functions, etc.)

**Pull request merged**